cmd/yace/main.go

```diff
@@ -140,18 +108,61 @@ func main() {
 		</html>`))
 	})
 
-	http.HandleFunc("/metrics", func(w http.ResponseWriter, r *http.Request) {
+	log.Fatal(http.ListenAndServe(*addr, nil))
+}
+
+type scraper struct {
+	cloudwatchSemaphore chan struct{}
+	tagSemaphore        chan struct{}
+	now                 time.Time
+	registry            *prometheus.Registry
+	// updateMutex         sync.Mutex // make sure updates are not causing race
+
+}
+
+func NewScraper() *scraper {
+	return &scraper{
+		cloudwatchSemaphore: make(chan struct{}, *cloudwatchConcurrency),
+		tagSemaphore:        make(chan struct{}, *tagConcurrency),
+		registry:            prometheus.NewRegistry(),
+	}
+}
+
+func (s *scraper) makeHandler() func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
 		if !(*decoupledScraping) {
-			newRegistry := prometheus.NewRegistry()
-			exporter.UpdateMetrics(config, newRegistry, now, *metricsPerQuery, *fips, *floatingTimeWindow, *labelsSnakeCase, cloudwatchSemaphore, tagSemaphore)
-			log.Debug("Metrics scraped.")
-			registry = newRegistry
+			s.scrape()
 		}
-		handler := promhttp.HandlerFor(registry, promhttp.HandlerOpts{
+		handler := promhttp.HandlerFor(s.registry, promhttp.HandlerOpts{
 			DisableCompression: false,
 		})
 		handler.ServeHTTP(w, r)
-	})
+	}
+}
 
-	log.Fatal(http.ListenAndServe(*addr, nil))
+func (s *scraper) decoupled() {
+	for {
+		log.Info("Starting scraping async")
+		// run scraping async
+		go s.scrape()
+
+		log.Info("Sleeping at regular sleep interval ", *scrapingInterval)
+		time.Sleep(time.Duration(*scrapingInterval) * time.Second)
+	}
```

Comment:
You could probably simplify this a bit:

```
ticker := time.NewTicker(time.Duration(*scrapingInterval) * time.Second)
defer ticker.Stop()
for {
  select {
  case <-ctx.Done():
    return // how do you handle shutdown atm?
  case <-ticker.C:
    if err := s.scrape(); err != nil {
      return err // something bad happened in scrape()
    }
  }
}
```

That way if `scrape()` takes a long time, it'll only ever run once in a program, but also gives you cancelation and ensures you only call it once every `ticker` interval.

The semaphore usage below is 100% a-okay, but can be elided if you move the control up to just `decoupled()`.